contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+                if (_orderBookSide) {
+                    return Array.from(_orderBookSide.deduped?.entries()).map((side) => [
+                        (side[0] / 10_000).toFixed(4),
+                        side[1],
+                    ]);
+                }
+            }
+            if (isBidSide) {
+                return [[0, 0]];
+            }
+            return [[69, 0]];
+        };
+
+        const getToB = (bids: LeafNode[], asks: LeafNode[]) => {
+            const _bids = orderBookSide(bids, true);
+            const _asks = orderBookSide(asks);
+            const tobAsk: number = Number(_asks[0][0]);
+            const tobBid: number = Number(_bids[0][0]);
+            return {
+                topAsk: tobAsk,
+                topBid: tobBid,
+            };
+        };
+
+        const getSpreadString = (bids: LeafNode[], asks: LeafNode[]) => {
+            const { topAsk, topBid } = getToB(bids, asks);
+            const spread: number = topAsk - topBid;
+            const spreadPercent: string = ((spread / topBid) * 100).toFixed(2);
+
+            return spread === topAsk
+                ? '∞ (100.00%)'
+                : `${spread.toFixed(2).toString()} (${spreadPercent}%)`;
+        };
+
+        if (markets) {
+            return {
+                passBidsProcessed: getSide(markets.passBids, true),
+                passAsksProcessed: getSide(markets.passAsks),
+                passBidsArray: orderBookSide(markets.passBids, true),
+                passAsksArray: orderBookSide(markets.passAsks),
+                failBidsProcessed: getSide(markets.failBids, true),
+                failAsksProcessed: getSide(markets.failAsks),
+                failBidsArray: orderBookSide(markets.failBids, true),
+                failAsksArray: orderBookSide(markets.failAsks),
+                passToB: getToB(markets.passBids, markets.passAsks),
+                failToB: getToB(markets.failBids, markets.failAsks),
+                passSpreadString: getSpreadString(markets.passBids, markets.passAsks),
+                failSpreadString: getSpreadString(markets.failBids, markets.failAsks),
+            };
+        }
+        return undefined;
+    }, [markets]);
+
+    const placeOrder = useCallback(
+        async (amount: number, price: number, limitOrder?: boolean, ask?: boolean, pass?: boolean) => {
+            if (!proposal || !markets) return;
+            const market = pass
+                ? { publicKey: proposal?.account.openbookPassMarket, account: markets?.pass }
+                : { publicKey: proposal?.account.openbookFailMarket, account: markets?.fail };
+            const placeTxs = await placeOrderTransactions(amount, price, market, limitOrder, ask, pass);
+
+            if (!placeTxs || !wallet.publicKey) {
+                return;
+            }
+
+            try {
+                setLoading(true);
+
+                await sender.send(placeTxs);
+                await fetchMarketsInfo();
```

Comment:
Given the above comment, I'm not entirely sure this is necessary, given the orderbooks update based on WS, maybe what we do here is check for WS connection and subscription, if it doesn't exist or is failed, then we fall back to a regular fetch. But MOST of the data in that fetchMarketsInfo() hasn't changed AFAIK.